### DISCUSSION TITLE
Harman AIR SDK release 33.1.1.300 mxml compile error:  mxml is of an unsupported type

### DISCUSSION THREAD

**Discussion by JoeZan77** [airsdk/Adobe-Runtime-Support#611] [Q&A]
We are trying to migrate our legacy application Flex/AIR application, but we receive a strange error even just by compiling a simple Hello
World application:
 
X1.mxml:
```
<?xml version="1.0" encoding="utf-8"?>
<s:Application xmlns:fx="http://ns.adobe.com/mxml/2009"
               xmlns:s="library://ns.adobe.com/flex/spark"
               xmlns:mx="library://ns.adobe.com/flex/mx" width="640" height="480">
    <s:layout>
        <s:BasicLayout />
    </s:layout>
    <s:Label text="Hello World!" horizontalCenter="0" verticalCenter="0" />
</s:Application>
```
 
This is the error:
./mxmlc X1.mxml
Loading configuration: /Users/g.zanella/Downloads/AIRSDK_MacOS (1)/frameworks/flex-config.xml
 
Error: /Users/g.zanella/Downloads/AIRSDK_MacOS (1)/bin/X1.mxml is of an unsupported type: mxml
 
with the legacy Apache Flex compiler all works fine:
 
/Applications/Adobe Flash Builder 4.6/sdks/4.16.0 AIR 31/bin/mxmlc X1.mxml
Create X1.swf
 
/Users/g.zanella/Downloads/AIRSDK_MacOS (1)/bin/mxmlc X1.mxml
Generates the above error
 
 
Thanks in advance for the reply

**Comment by aritgithub** [comment]
Similar errors in my applications with **AIR SDK 33** that I have not had time to investigate but that with the legacy version (**AIR SDK 32**) compile and work correctly.

**Comment by ajwfrost** [comment]
I've had previous issues in the past (with older AIR SDKs from Adobe) and I believe it's related to the Java settings (JAVA_HOME values, path information, and installed versions of Java). But it would be good to get to the bottom of exactly what causes that output..

Note that if you are using Flex, you should be using the Flex SDK and then overlaying the AIR SDK that's specifically for Flex developers - this package shouldn't then include any compilers, so the message there I think comes from the Flex compiler but perhaps is caused by an exception happening that's due to the other SDK files..

**Comment by JanFW** [comment]
I have had this issue before, too, and it is most likely a SDK overlaying issue, as Andrew suspects.
Searching for _"mxml is of an unsupported type: mxml"_ will reveal several threads pointing to an overlay issues where the non-Flex AIR SDK was used for overlaying.

**Comment by AndreaFilippi** [comment]
I have the same issue compiling with "Full AIR SDK with new ActionScript Compiler" version 33.1.1.476:
    _INFO: Couldn't access Thread.inheritableThreadLocals. Reference finalizer threads will inherit thread local values.
    Loading configuration: C:\AIRSDK_Windows\frameworks\air-config.xml
    Error: <file>.mxml is of an unsupported type: mxml_
Any update on this problem?
I do not get errors with the Apache compiler version 4.16.1

**Reply by ajwfrost** [reply]
If you're using the "new ActionScript Compiler" version, then it doesn't support MXML, it's just got the AS3 capabilities in it. To compile MXML, you'll need to download the Flex SDK (which contains mxmlc.jar) and then overlay on top of this the AIR SDK for Flex developers.

I'm not quite sure why it's set up like this - historical reasons I guess, but the driving apps are the scripts/batch files in the 'bin' folders which are different and could perhaps be merged together and execute depending on which file type is passed in...

i.e. MXMLC.bat from Flex:
```java  ...  -jar "%FLEX_HOME%\lib\mxmlc.jar" +flexlib="%FLEX_HOME%\frameworks" %*```
vs MXMLC.bat from AIR:
```java ... -Dflexlib="%AIR_SDK_HOME%\frameworks" -jar "%AIR_SDK_HOME%\lib\mxmlc-cli.jar" %*```

**Reply by ciacob** [reply]
Right, so you should be aware that the AIRSDK manager has a function to overlay an existing _regular_ AIR SDK over a Flex folder, which is, in my opinion, bogus and misleading. Since the "AIR SDK for Flex developers" is different than the regular AIR SDK, the fact that the manger makes no differentiation, and blindly offers to do the overlay **even where it makes no sense**, is just wrong.

I just though I would point that out.

<img width="1805" height="311" alt="image" src="https://github.com/user-attachments/assets/638b5c68-ccee-4750-95a9-755600c7ee18" />

In my case, I had just used that bogus function, and ended up with an unusable overlayed SDK I eventually had to delete and redo manually.

**Reply by ajwfrost** [reply]
Hmm... that mechanism should only copy over the relevant parts of the AIR SDK, leaving the original Flex compiler still in place. The AIR SDKs that you download within the AIR SDK Manager are provided in a set of different zip files, the overlay happens by taking the platform-specific and generic zip files but leaving out the AIR-specific compiler/etc packages (i.e. the difference between the 'normal' AIR SDK, and the 'for Flex developers' AIR SDK. 

So this mechanism should work! If you still have the folder which isn't working, can you do something like `mxmlc -version` to check what compiler you've ended up with?

thanks

**Reply by ciacob** [reply]
So, was I supposed to use it on top of a folder that was already overlayed, i.e., on top of an already working "Flex + older AIR version"? Sorry, I went for a clean Flex folder, and that didn't work.

Also, I don't have that folder anymore. As it turns out, I always go for cleaning/resetting things... 🤷

**Reply by ajwfrost** [reply]
You should be able to do either .. but it sounded like something went wrong. We've tried it here and the copying worked okay, but if you encounter the issue again then if you can check the compiler version, it may help us work out what went wrong.
The alternative is if you want the monolithic zip file for the AIR SDK for Flex i.e. without compilers, then it's still available via the "Getting Started" part of https://airsdk.dev

thanks

**Comment by ajwfrost** [comment]
Hi all
A number of folk have still been having problems with getting a working "Flex + AIR" SDK together, so we've created a quick utility that will take a Flex SDK (whether or not it's already been overlaid), plus an AIR SDK, and will then create a 'merged' folder that overlays the AIR components onto the Flex SDK.
It handles a couple of things
- ensures that it is the right kind of AIR SDK!
- ensures the 'MainWindow.nib' issues aren't a problem, when copying from different AIR SDK versions (sometimes these are files, sometimes they're folders)

We've just packaged this as an .air file so you would need to have the shared runtime installed first (https://airsdk.harman.com/runtime) and then you can run this, search for/select the input folders, select a fresh folder for the output, and hit 'merge'.... (you'll need to unzip the below file to get at the .air file!)
[MergeAIRandFlex.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/8184807/MergeAIRandFlex.zip)

thanks

**Comment by nicopowa** [comment]
Playing with my 10 years old code :)
1 hour trying to install AIR50 + FlexSDK on W10, not working
- Downloaded 
AIR 32 + Compiler from Adobe : https://helpx.adobe.com/air/kb/archived-air-sdk-version.html
Flex SDK 4.6 : https://stackoverflow.com/a/21269691/3089199
Latest Flex SDK : https://airsdk.harman.com/download
- Read [this page](https://github.com/BowlerHatLLC/vscode-as3mxml/wiki/Install-the-ActionScript-and-MXML-extension-for-Visual-Studio-Code)
- Added AIRSDK 32 directory to PATH
- Dropped latest Flex SDK in 4.6 directory
- Magic ! Debugging in VSCode
- Bugs everywhere, but it's running

**Comment by hnuecke** [comment]
Sorry jumping on this thread with probably a not fully related question!

I am still maintaining an old Flex/AIR App. Which runs perfectly fine on Windows11.
But when trying to "re-animate" my MAC version, I fail desperately!
I assume this is due to an incompatibility of the Flex SDK and my macOS version, and am looking for either a confirmation (then I would give up my MAC support) or a suggestion how to get it working.

Here's what I did and have:
Mac mini late 2012 with Catalina 10.15.7
Folder with Apache Flex 4.16.1 as copy of working folder with an Air 33 merge, where I merged in the latest AIR52 Flex SDK.
I get a compile error that I interpret the way that a flex component is missing:
     Build failed: Could not find or load main class flex2.tools.Mxmlc

I tried running the Apache Flex installer, which immediately failed; probably also for not supporting the latest macOS version:
    ?info=Error #2031

If anybody has an idea or suggestion, that would be highly welcome!!
Regards
Hans


    
![preview1](https://user-images.githubusercontent.com/3041318/227267744-c0a0b70b-5278-4353-981f-73828352788c.png)
![preview2](https://user-images.githubusercontent.com/3041318/227268065-837e4a45-9000-4eb1-baad-37f1c38b15fc.png)
![preview3](https://user-images.githubusercontent.com/3041318/227268591-5e5d26d3-d4d1-44b8-8da9-06faabe1c05e.png)

**Reply by ajwfrost** [reply]
Theoretically there's no reason why this wouldn't work.. the error you describe sounds like something has gone wrong in the merging/overlay of the AIR tools over the top of the Flex SDK. MacOS has some peculiarities when it comes to folder/file overwriting and merging.... So the best thing to do would be:
1) create a fresh copy of what you know you had working earlier (Flex 4.16.1 etc)
2) download the AIR Flex-compatible SDK and expand that into a new location
3) copy the AIR stuff over the Flex stuff with `cp -Rf ~/tmp/AIRSDK_Flex_MacOS/* /path-to-empty-FLEXSDK-directory`
Note the capital 'R' to ensure links get copied properly...

The Flex installer I think no longer works properly; if you do download the Flex SDK from scratch then you may have to update some variables in the flex config XML files such as `{airHome}` which are what that installer app is meant to resolve (just with the SDK base folder).

Hope that helps....
